libraries/typescript/Microsoft.Bot.Builder.Adapters.WeChat/src/weChatAdapter.ts

```diff
@@ -0,0 +1,373 @@
+                    const responses = context.turnState.get(this.TurnResponseKey);
+                    if (responses.has(key)) {
+                        responses.get(key).push(activity);
+                    } else {
+                        responses.set(key, [activity]);
+                    }
+                    break;
+                default:
+                    break;
+            }
+            const resourceResponse: ResourceResponse = {
+                id: activity.id || ''
+            };
+            resourceResponses.push(resourceResponse);
+        }
+        return resourceResponses;
+    }
+
+    /**
+     * Process the request from WeChat.
+     * @param req The request sent from WeChat.
+     * @param res Http response object of current request.
+     * @param logic A function handler that will be called to perform the bots logic after the the adapters middleware has been run.
+     * @param secretInfo Secret info for verify the request.
+     * @param passiveResponse If using passvice response mode, if set to true, user can only get one reply.
+     * @returns Process activity result.
+     */
+    public async processActivity(req: WebRequest, res: WebResponse, logic: (context: TurnContext) => Promise<any>, secretInfo: SecretInfo, passiveResponse: boolean): Promise<void> {
+        if (!req) {
+            throw new Error(`ArgumentNullException - Request is invalid.`);
+        }
+
+        if (!res) {
+            throw new Error(`ArgumentNullException - Response is invalid.`);
+        }
+
+        if (!logic) {
+            throw new Error(`ArgumentNullException - Bot logic is invalid.`);
+        }
+
+        if (!secretInfo) {
+            throw new Error(`ArgumentNullException - Secret information is invalid.`);
+        }
+
+        if (!VerificationHelper.VerifySignature(secretInfo.Signature, secretInfo.Timestamp, secretInfo.Nonce, this.settings.Token)) {
+            throw new Error('UnauthorizedAccessException - Signature verification failed');
+        }
+
+        secretInfo.Token = this.settings.Token;
+        secretInfo.EncodingAesKey = this.settings.EncodingAESKey;
+        secretInfo.AppId = this.settings.AppId;
+        const weChatRequest = await parseRequest(req, secretInfo);
+        if (!passiveResponse) {
+            this.ProcessWeChatRequest(weChatRequest, logic, passiveResponse);
+            // Return status
+            res.status(200);
+            res.end();
+        } else {
+            // TODO: Passive reply
+        }
+    }
+
+
+    /**
+     * Get the respone from bot for the wechat request.
+     * @param activities List of bot activities.
+     * @param openId User's open id from WeChat.
+     * @param passiveResponse If using passvice response mode, if set to true, user can only get one reply.
+     * @returns Bot response message.
+     */
+    private async ProcessBotResponse(activities: Activity[], openId: string, passiveResponse: boolean): Promise<any> {
```

Comment:
>ProcessBotResponse [](start = 18, length = 18)

Just curious. why do you wait and send all of the responses at the end of the turn versus sending them as the bot calls sendActivities() ?  While it's valid to do it this the issue is you now can't tell the bots logic when a message delivery failed.

Reply:
WeChat we need a lot of time to map the activity to WeChat message format which include uploading media, like image and video, etc.(WeChat only accept media link in it’s domain). This cost a lot of time, but WeChat required to return 200 in 5 second, so in c# version, we just queue the reply activities and process them in the background and I followed it in IS version

Reply:
WeChat we need a lot of time to map the activity to WeChat message format which include uploading media, like image and video, etc.(WeChat only accept media link in it’s domain). This cost a lot of time, but WeChat required to return 200 in 5 second, so in c# version, we just queue the reply activities and process them in the background and I followed it in IS version